File: pkg/tanka/export.go
package tanka

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"io/fs"
	"log"
	"os"
	"path/filepath"
	"strings"
	"text/template"

	"github.com/Masterminds/sprig/v3"
	"github.com/pkg/errors"
	"k8s.io/apimachinery/pkg/labels"

	"github.com/grafana/tanka/pkg/kubernetes/manifest"
	"github.com/grafana/tanka/pkg/spec/v1alpha1"
)

// BelRune is a string of the Ascii character BEL which made computers ring in ancient times
// We use it as "magic" char for the subfolder creation as it is a non printable character and thereby will never be
// in a valid filepath by accident. Only when we include it.
const BelRune = string(rune(7))

// When exporting manifests to files, it becomes increasingly hard to map manifests back to its environment, this file
// can be used to map the files back to their environment. This is aimed to be used by CI/CD but can also be used for
// debugging purposes.
const manifestFile = "manifest.json"

type ExportMergeStrategy string

const (
	ExportMergeStrategyNone          ExportMergeStrategy = ""
	ExportMergeStrategyFailConflicts ExportMergeStrategy = "fail-on-conflicts"
	ExportMergeStrategyReplaceEnvs   ExportMergeStrategy = "replace-envs"
)

// ExportEnvOpts specify options on how to export environments
type ExportEnvOpts struct {
	// formatting the filename based on the exported Kubernetes manifest
	Format string
	// extension of the filename
	Extension string
	// optional: options to parse Jsonnet
	Opts Opts
	// optional: filter environments based on labels
	Selector labels.Selector
	// optional: number of environments to process in parallel
	Parallelism int

	// What to do when exporting to an existing directory
	// - none: fail when directory is not empty
	// - fail-on-conflicts: fail when an exported file already exists
	// - replace-envs: delete files previously exported by the targeted envs and re-export them
	MergeStrategy ExportMergeStrategy
}

func ExportEnvironments(envs []*v1alpha1.Environment, to string, opts *ExportEnvOpts) error {
	// Keep track of which file maps to which environment
	fileToEnv := map[string]string{}

	// dir must be empty
	empty, err := dirEmpty(to)
	if err != nil {
		return fmt.Errorf("checking target dir: %s", err)
	}
	if !empty && opts.MergeStrategy == ExportMergeStrategyNone {
		return fmt.Errorf("output dir `%s` not empty. Pass a different --merge-strategy to ignore this", to)
	}

	// delete files previously exported by the targeted envs.
	if opts.MergeStrategy == ExportMergeStrategyReplaceEnvs {
		if err := deletePreviouslyExportedManifests(to, envs); err != nil {
			return fmt.Errorf("deleting previously exported manifests: %w", err)
		}
	}

	// get all environments for paths
	loadedEnvs, err := parallelLoadEnvironments(envs, parallelOpts{
		Opts:        opts.Opts,
		Selector:    opts.Selector,
		Parallelism: opts.Parallelism,
	})
	if err != nil {
		return err
	}

	for _, env := range loadedEnvs {
		// get the manifests
		loaded, err := LoadManifests(env, opts.Opts.Filters)
		if err != nil {
			return err
		}

		env := loaded.Env
		res := loaded.Resources

		// create raw manifest version of env for templating
		env.Data = nil
		raw, err := json.Marshal(env)
		if err != nil {
			return err
		}
		var menv manifest.Manifest
		if err := json.Unmarshal(raw, &menv); err != nil {
			return err
		}

		// create template
		manifestTemplate, err := createTemplate(opts.Format, menv)
		if err != nil {
			return fmt.Errorf("parsing format: %s", err)
		}

		// write each to a file
		for _, m := range res {
			// apply template
			name, err := applyTemplate(manifestTemplate, m)
			if err != nil {
				return fmt.Errorf("executing name template: %w", err)
			}

			// Create all subfolders in path
			relpath := name + "." + opts.Extension
			path := filepath.Join(to, relpath)

			fileToEnv[relpath] = env.Metadata.Namespace

			// Abort if already exists
			if exists, err := fileExists(path); err != nil {
				return err
			} else if exists {
				return fmt.Errorf("file '%s' already exists. Aborting", path)
			}

			// Write manifest
			data := m.String()
			if err := writeExportFile(path, []byte(data)); err != nil {
				return err
			}
		}
	}

	return exportManifestFile(to, fileToEnv, nil)
}

func fileExists(name string) (bool, error) {
	_, err := os.Stat(name)
	if os.IsNotExist(err) {
		return false, nil
	}
	if err != nil {
		return false, err
	}
	return true, nil
}

func dirEmpty(dir string) (bool, error) {
	f, err := os.Open(dir)
	if os.IsNotExist(err) {
		return true, os.MkdirAll(dir, os.ModePerm)
	} else if err != nil {
		return false, err
	}
	defer f.Close()

	_, err = f.Readdirnames(1)
	if err == io.EOF {
		return true, nil
	}
	return false, err
}

func deletePreviouslyExportedManifests(path string, envs []*v1alpha1.Environment) error {
	fileToEnvMap := make(map[string]string)

	manifestFilePath := filepath.Join(path, manifestFile)
	manifestContent, err := os.ReadFile(manifestFilePath)
	if err != nil && errors.Is(err, fs.ErrNotExist) {
		log.Printf("Warning: No manifest file found at %s, skipping deletion of previously exported manifests\n", manifestFilePath)
		return nil
	} else if err != nil {
		return err
	}

	if err := json.Unmarshal(manifestContent, &fileToEnvMap); err != nil {
		return err
	}

	envNames := make(map[string]struct{})
	for _, env := range envs {
		envNames[env.Metadata.Namespace] = struct{}{}
	}

	var deletedManifestKeys []string
	for exportedManifest, manifestEnv := range fileToEnvMap {
		if _, ok := envNames[manifestEnv]; ok {
			deletedManifestKeys = append(deletedManifestKeys, exportedManifest)
			if err := os.Remove(filepath.Join(path, exportedManifest)); err != nil {
				return err
			}
		}
	}

	return exportManifestFile(path, nil, deletedManifestKeys)
}

// exportManifestFile writes a manifest file that maps the exported files to their environment.
// If the file already exists, the new entries will be merged with the existing ones.
func exportManifestFile(path string, newFileToEnvMap map[string]string, deletedKeys []string) error {
	if len(newFileToEnvMap) == 0 && len(deletedKeys) == 0 {
		return nil
	}

	manifestFilePath := filepath.Join(path, manifestFile)
	manifestContent, err := os.ReadFile(manifestFilePath)
	currentFileToEnvMap := make(map[string]string)
	if err != nil && !errors.Is(err, fs.ErrNotExist) {
		return fmt.Errorf("reading existing manifest file: %w", err)
	} else if err == nil {
		if err := json.Unmarshal(manifestContent, &currentFileToEnvMap); err != nil {
			return fmt.Errorf("unmarshalling existing manifest file: %w", err)
		}
	}

	for k, v := range newFileToEnvMap {
		currentFileToEnvMap[k] = v
	}
	for _, k := range deletedKeys {
		delete(currentFileToEnvMap, k)
	}

	// Write manifest file
	data, err := json.MarshalIndent(currentFileToEnvMap, "", "    ")
	if err != nil {
		return fmt.Errorf("marshalling manifest file: %w", err)
	}

	return writeExportFile(manifestFilePath, data)
}

func writeExportFile(path string, data []byte) error {
	if err := os.MkdirAll(filepath.Dir(path), 0700); err != nil {
		return fmt.Errorf("creating filepath '%s': %s", filepath.Dir(path), err)
	}

	return os.WriteFile(path, data, 0644)
}

func createTemplate(format string, env manifest.Manifest) (*template.Template, error) {
	// Replace all os.path separators in string with BelRune for creating subfolders
	replaceFormat := replaceTmplText(format, string(os.PathSeparator), BelRune)

	envMap := template.FuncMap{"env": func() manifest.Manifest { return env }}

	template, err := template.New("").
		Funcs(sprig.TxtFuncMap()). // register Masterminds/sprig
		Funcs(envMap).             // register environment mapping
		Parse(replaceFormat)       // parse template
	if err != nil {
		return nil, err
	}
	return template, nil
}

func replaceTmplText(s, old, new string) string {
	parts := []string{}
	l := strings.Index(s, "{{")
	r := strings.Index(s, "}}") + 2

	for l != -1 && l < r {
		// replace only in text between template action blocks
		text := strings.ReplaceAll(s[:l], old, new)
		action := s[l:r]
		parts = append(parts, text, action)
		s = s[r:]
		l = strings.Index(s, "{{")
		r = strings.Index(s, "}}") + 2
	}
	parts = append(parts, strings.ReplaceAll(s, old, new))
	return strings.Join(parts, "")
}

func applyTemplate(template *template.Template, m manifest.Manifest) (path string, err error) {
	buf := bytes.Buffer{}
	if err := template.Execute(&buf, m); err != nil {
		return "", err
	}

	// Replace all os.path separators in string in order to not accidentally create subfolders
	path = strings.ReplaceAll(buf.String(), string(os.PathSeparator), "-")
	// Replace the BEL character inserted with a path separator again in order to create a subfolder
	path = strings.ReplaceAll(path, BelRune, string(os.PathSeparator))

	return path, nil
}

File: rust-export/src/export.rs
use anyhow::{anyhow, Context, Result};
use log::{info, warn};
use rayon::prelude::*;
use regex::Regex;
use std::collections::HashMap;
use std::fs;
use std::io::ErrorKind;
use std::path::{Path, PathBuf};
use walkdir::WalkDir;

use crate::environment::LoadedEnvironment;
use crate::jsonnet::{load_environment, load_manifests, list_environments};
use crate::template::TemplateEngine;

const MANIFEST_FILE: &str = "manifest.json";

#[derive(Debug, Clone)]
pub enum ExportMergeStrategy {
    None,
    FailOnConflicts,
    ReplaceEnvs,
}

#[derive(Debug, Clone)]
pub struct ExportOptions {
    pub format: String,
    pub extension: String,
    pub parallelism: usize,
    pub cache_path: Option<PathBuf>,
    pub cache_path_regexes: Vec<Regex>,
    pub merge_strategy: ExportMergeStrategy,
    pub merge_deleted_envs: Vec<String>,
    pub skip_manifest: bool,
    pub targets: Vec<String>,
    pub name: Option<String>,
    pub selector: Option<String>,
}

pub async fn export_environments(
    output_dir: PathBuf,
    paths: Vec<PathBuf>,
    recursive: bool,
    opts: ExportOptions,
) -> Result<()> {
    // Check if directory is empty
    let empty = is_dir_empty(&output_dir)?;
    if !empty && matches!(opts.merge_strategy, ExportMergeStrategy::None) {
        return Err(anyhow!(
            "output dir '{}' not empty. Pass a different --merge-strategy to ignore this",
            output_dir.display()
        ));
    }

    // Find environments
    let environments = if recursive {
        find_environments_recursive(&paths, &opts)?
    } else {
        if paths.len() > 1 {
            return Err(anyhow!(
                "recursive flag is required when exporting multiple environments"
            ));
        }
        vec![load_environment(&paths[0])?]
    };

    info!("Found {} environment(s)", environments.len());

    // Delete previously exported manifests if using replace strategy
    if matches!(opts.merge_strategy, ExportMergeStrategy::ReplaceEnvs) {
        delete_previously_exported_manifests_from_envs(&output_dir, &environments, opts.skip_manifest)?;
    }

    // Delete manifests from deleted environments
    if !opts.merge_deleted_envs.is_empty() {
        delete_previously_exported_manifests(&output_dir, &opts.merge_deleted_envs, opts.skip_manifest)?;
    }

    // Export environments in parallel
    let file_to_env: HashMap<String, String> = environments
        .par_iter()
        .map(|env| export_single_environment(env, &output_dir, &opts))
        .collect::<Result<Vec<_>>>()?
        .into_iter()
        .flatten()
        .collect();

    // Write manifest file
    if !opts.skip_manifest {
        export_manifest_file(&output_dir, &file_to_env, &[])?;
    }

    info!("Successfully exported {} environment(s)", environments.len());
    Ok(())
}

fn find_environments_recursive(paths: &[PathBuf], opts: &ExportOptions) -> Result<Vec<LoadedEnvironment>> {
    let mut environments = Vec::new();

    for path in paths {
        for entry in WalkDir::new(path)
            .follow_links(true)
            .into_iter()
            .filter_map(|e| e.ok())
        {
            if entry.file_name() == "main.jsonnet" {
                // Use list_environments which can handle files with multiple environments
                match list_environments(entry.path()) {
                    Ok(envs) => {
                        for env in envs {
                            // Filter by name if specified
                            if let Some(ref name) = opts.name {
                                if let Some(ref env_name) = env.environment.metadata.name {
                                    if env_name != name {
                                        continue;
                                    }
                                } else {
                                    continue;
                                }
                            }

                            // TODO: Filter by selector if specified
                            environments.push(env);
                        }
                    }
                    Err(e) => {
                        warn!("Failed to load environment at {:?}: {}", entry.path(), e);
                    }
                }
            }
        }
    }

    Ok(environments)
}

fn export_single_environment(
    env: &LoadedEnvironment,
    output_dir: &Path,
    opts: &ExportOptions,
) -> Result<HashMap<String, String>> {
    info!("Exporting environment: {:?}", env.path);

    let manifests = load_manifests(env)?;
    
    if manifests.is_empty() {
        info!("No manifests found in environment: {:?}", env.path);
        return Ok(HashMap::new());
    }

    let template = TemplateEngine::new(&opts.format, &env.environment)?;
    let mut file_to_env = HashMap::new();

    for manifest in manifests {
        let name = template.apply(&manifest, &env.environment)?;
        let relpath = format!("{}.{}", name, opts.extension);
        let path = output_dir.join(&relpath);

        // Check if file exists (for fail-on-conflicts mode)
        if path.exists() && matches!(opts.merge_strategy, ExportMergeStrategy::FailOnConflicts) {
            return Err(anyhow!("file '{}' already exists. Aborting", path.display()));
        }

        // Get environment namespace
        let env_namespace = env.environment.metadata.namespace.as_deref().unwrap_or("default");
        file_to_env.insert(relpath, env_namespace.to_string());

        // Write manifest
        let yaml = manifest.to_yaml()?;
        write_export_file(&path, yaml.as_bytes())?;
    }

    Ok(file_to_env)
}

fn is_dir_empty(path: &Path) -> Result<bool> {
    match fs::read_dir(path) {
        Ok(mut entries) => Ok(entries.next().is_none()),
        Err(e) if e.kind() == ErrorKind::NotFound => {
            fs::create_dir_all(path)?;
            Ok(true)
        }
        Err(e) => Err(e.into()),
    }
}

fn write_export_file(path: &Path, data: &[u8]) -> Result<()> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)
            .with_context(|| format!("Failed to create directory: {:?}", parent))?;
    }

    fs::write(path, data)
        .with_context(|| format!("Failed to write file: {:?}", path))?;

    Ok(())
}

fn delete_previously_exported_manifests_from_envs(
    path: &Path,
    envs: &[LoadedEnvironment],
    skip_manifest: bool,
) -> Result<()> {
    let env_names: Vec<String> = envs
        .iter()
        .filter_map(|e| e.environment.metadata.namespace.clone())
        .collect();
    
    delete_previously_exported_manifests(path, &env_names, skip_manifest)
}

fn delete_previously_exported_manifests(
    path: &Path,
    tanka_env_names: &[String],
    skip_manifest: bool,
) -> Result<()> {
    if tanka_env_names.is_empty() {
        return Ok(());
    }

    let env_names_set: HashMap<&str, ()> = tanka_env_names
        .iter()
        .map(|s| (s.as_str(), ()))
        .collect();

    let manifest_file_path = path.join(MANIFEST_FILE);
    let manifest_content = match fs::read_to_string(&manifest_file_path) {
        Ok(content) => content,
        Err(e) if e.kind() == ErrorKind::NotFound => {
            warn!(
                "No manifest file found at {:?}, skipping deletion of previously exported manifests",
                manifest_file_path
            );
            return Ok(());
        }
        Err(e) => return Err(e.into()),
    };

    let file_to_env_map: HashMap<String, String> = serde_json::from_str(&manifest_content)
        .context("Failed to parse manifest file")?;

    let mut deleted_manifest_keys = Vec::new();
    for (exported_manifest, manifest_env) in &file_to_env_map {
        if env_names_set.contains_key(manifest_env.as_str()) {
            deleted_manifest_keys.push(exported_manifest.clone());
            let file_path = path.join(exported_manifest);
            if let Err(e) = fs::remove_file(&file_path) {
                if e.kind() != ErrorKind::NotFound {
                    return Err(e.into());
                }
            }
        }
    }

    // Update manifest file
    if !skip_manifest {
        export_manifest_file(path, &HashMap::new(), &deleted_manifest_keys)?;
    }

    Ok(())
}

fn export_manifest_file(
    path: &Path,
    new_file_to_env_map: &HashMap<String, String>,
    deleted_keys: &[String],
) -> Result<()> {
    if new_file_to_env_map.is_empty() && deleted_keys.is_empty() {
        return Ok(());
    }

    let manifest_file_path = path.join(MANIFEST_FILE);
    let mut current_file_to_env_map: HashMap<String, String> = match fs::read_to_string(&manifest_file_path) {
        Ok(content) => serde_json::from_str(&content)?,
        Err(e) if e.kind() == ErrorKind::NotFound => HashMap::new(),
        Err(e) => return Err(e.into()),
    };

    // Merge new entries
    for (k, v) in new_file_to_env_map {
        current_file_to_env_map.insert(k.clone(), v.clone());
    }

    // Delete removed entries
    for k in deleted_keys {
        current_file_to_env_map.remove(k);
    }

    // Write manifest file
    let data = serde_json::to_string_pretty(&current_file_to_env_map)?;
    write_export_file(&manifest_file_path, data.as_bytes())?;

    Ok(())
}

